src/fidesops/service/saas_request/saas_request_override_factory.py

```diff
@@ -0,0 +1,159 @@
+import logging
+from enum import Enum
+from inspect import Signature, signature
+from typing import Callable, Dict, List, Union
+
+from fidesops.common_exceptions import (
+    InvalidSaaSRequestOverrideException,
+    NoSuchSaaSRequestOverrideException,
+)
+from fidesops.util.collection_util import Row
+
+logger = logging.getLogger(__name__)
+
+# at some point this should likely be formalized more centrally...
+class SaaSRequestType(Enum):
+    """
+    An `Enum` containing the different possible types of SaaS requests
+    """
+
+    READ = "read'"
+    UPDATE = "update"
+    DATA_PROTECTION_REQUEST = "data_protection_request"
+    DELETE = "delete"
+
+
+class SaaSRequestOverrideFactory:
+    """
+    Factory class responsible for registering, maintaining, and providing
+    user-defined functions that act as overrides to SaaS request execution
+    """
+
+    registry: Dict[
+        SaaSRequestType, Dict[str, Callable[..., Union[List[Row], int]]]
+    ] = {}
+    valid_overrides: Dict[SaaSRequestType, str] = {}
+
+    # initialize each request type's inner dicts with an empty dict
+    for request_type in SaaSRequestType:
+        registry[request_type] = {}
+        valid_overrides[request_type] = ""
+
+    @classmethod
+    def register(
+        cls, name: str, request_types: List[SaaSRequestType]
+    ) -> Callable[
+        [Callable[..., Union[List[Row], int]]], Callable[..., Union[List[Row], int]]
+    ]:
+        """
+        Decorator to register the custom-implemented saas request override
+        with the given name.
+        """
+
+        if isinstance(request_types, SaaSRequestType):
+            request_types = [request_types]
+        elif not request_types:
+            raise TypeError(
+                "At least one SaaSRequestType must be specified when registering SaaS request override function {name}"
+            )
+
+        def wrapper(
+            override_function: Callable[..., Union[List[Row], int]],
+        ) -> Callable[..., Union[List[Row], int]]:
+            for request_type in request_types:
+                logger.debug(
+                    f"Registering new SaaS request override function '{override_function}' under name '{name}' for SaaSRequestType {request_type}"
```

Comment:
Does a Callable turn into a user-friendly string? I tried it locally and got `<function add at 0x102722ca0>`. Maybe we can use `override_function.__name__`?